### PR TITLE
[runtime] Don't use global operators new and delete. Don't allow C++ exports.

### PR DESCRIPTION
**Pull request by gparker42**
Global operators new and delete can be overridden by app implementations,
but such implementations are historically untrustworthy. We instead use
our own implementations in libswiftCore. We share no C++ objects with
other code so we don't have interoperability problems with mismatched
operator implementations.

Test stdlib/RuntimeSymbols.swift verifies that operators new and delete
were not imported from libc++ and verifies that libswiftCore does not
itself export any C++-mangled symbols.

rdar://16891362

### CODE REVIEW

**Comment by gparker42** [comment]
@swift-ci please test

**Comment by gparker42** [comment]
@swift-ci please test

File: validation-test/stdlib/RuntimeSymbols.swift [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # Verify symbols from libswiftCore.dylib.
 
-UNSUPPORTED: OS=linux
+UNSUPPORTED: OS=linux-gnu
```
Comment (jrose-apple) [inline validation-test/stdlib/RuntimeSymbols.swift]: How are we going to verify this on Linux, then?

Comment (gparker42) [inline validation-test/stdlib/RuntimeSymbols.swift]: For now we won't. I don't think there is a common set of parameters to `nm` that do what we want (`-U` vs `--defined-only`), so we would need a second test with a Linux-compatible `nm` command.

Comment (gparker42) [inline validation-test/stdlib/RuntimeSymbols.swift]: The clean way is to add %target-nm-imports and %target-nm-exports to lit.cfg. I'll do that.

**Comment by gparker42** [comment]
@swift-ci please smoke test

File: validation-test/stdlib/RuntimeSymbols.swift (hunk 1 above)
Comment (jrose-apple) [inline validation-test/stdlib/RuntimeSymbols.swift]: I'm not sure that's actually an improvement. We'll never use it anywhere else. It's probably okay to just have two tests.

**Comment by gparker42** [comment]
@swift-ci please smoke test

**Comment by jrose-apple** [comment]
Reposting since it got folded away: I'm not sure having a lit substitution is actually an improvement. We'll never use it anywhere else. It's probably okay to just have two tests.

File: include/swift/Runtime/OperatorNew.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+// Never use the global definitions of operators new and delete.
+// The app may replace them and we don't trust their implementations.
+
+#include <new>
+#include <stdlib.h>
+
+#define SWIFT_PRIVATE_INLINE \
+  __private_extern__ inline __attribute__((always_inline))
+
```
Comment (hughbe) [inline include/swift/Runtime/OperatorNew.h]: Can we use The LLVM attribute for this?

Comment (hughbe) [inline include/swift/Runtime/OperatorNew.h]: ...or the swift attribute if LLVM doesnt define anything like this :)

Comment (gparker42) [inline include/swift/Runtime/OperatorNew.h]: I don't see any existing wrappers around __private_extern__ in llvm or swift. It's possible that __private_extern__ is unnecessary given always_inline.

**Comment by gparker42** [comment]
This won't be the last test of symbol exports. It's possible that platform differences will force us to use multiple tests in the future anyway, but I'd like to avoid that for as long as we can.

**Comment by gparker42** [comment]
This isn't right. The demangler code is compiled into three different places (libswiftCore, libswiftRemoteMirror, and several tools). Using the custom inline operators unconditionally in that code is bad because only libswiftCore is going to cooperate with that.

**Pull request closed without merging**